agent/stats/task.go

```diff
@@ -0,0 +1,241 @@
+	PopulateNIDeviceList(containerPID string) ([]string, error)
+}
+
+type TaskStatsStruct struct {
+	netlinkClient netlinkwrapper.NetLink
+	nsclient nswrapper.NS
+}
+
+
+func newStatsTaskContainer(taskARN string, containerPID string, numberOfContainers int,
+	resolver resolver.ContainerMetadataResolver) (*StatsTask, error) {
+	nsAgent := nswrapper.NewNS()
+	netlinkclient := netlinkwrapper.New()
+
+	taskStatsStruct := &TaskStatsStruct{
+		netlinkClient: netlinkclient,
+		nsclient:      nsAgent,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	return &StatsTask{
+		TaskMetadata: &TaskMetadata{
+			TaskArn:          taskARN,
+			ContainerPID:     containerPID,
+			NumberContainers: numberOfContainers,
+		},
+		Ctx:      ctx,
+		Cancel:   cancel,
+		Resolver: resolver,
+		client:   taskStatsStruct,
+	}, nil
+
+}
+
+func (task *StatsTask) StartStatsCollectionTask() {
+	queueSize := int(config.DefaultContainerMetricsPublishInterval.Seconds() * 4)
+	task.StatsQueue = NewQueue(queueSize)
+	task.StatsQueue.Reset()
+	go task.collect()
+}
+
+func (task *StatsTask) StopStatsCollectionTask() {
+	task.Cancel()
+}
+
+func (taskStat *StatsTask) collect() {
+	taskArn := taskStat.TaskMetadata.TaskArn
+	for {
+		select {
+		case <-taskStat.Ctx.Done():
+			seelog.Debugf("Stopping stats collection for taskStat %s", taskArn)
+			return
+		default:
+			err := taskStat.processStatsStream()
+			if err != nil {
+				seelog.Debugf("Error querying stats for task %s: %v", taskArn, err)
+			}
+			// We were disconnected from the stats stream.
+			// Check if the task is terminal. If it is, stop collecting metrics.
+			terminal, err := taskStat.terminal()
+			if err != nil {
+				// Error determining if the task is terminal. clean-up anyway.
+				seelog.Warnf("Error determining if the task %s is terminal, stopping stats collection: %v",
+					taskArn, err)
+				taskStat.StopStatsCollectionTask()
+			} else if terminal {
+				seelog.Infof("Task %s is terminal, stopping stats collection", taskArn)
+				taskStat.StopStatsCollectionTask()
+			}
+		}
+	}
+}
+
+func (taskStat *StatsTask) processStatsStream() error {
+	taskArn := taskStat.TaskMetadata.TaskArn
+	if len(taskStat.TaskMetadata.DeviceName) == 0 {
+		var err error
+		taskStat.TaskMetadata.DeviceName, err = taskStat.client.PopulateNIDeviceList(taskStat.TaskMetadata.ContainerPID)
+		if err != nil {
+			return err
+		}
+	}
+	awsvpcNetworkStats, errC := taskStat.client.GetAWSVPCNetworkStats(taskStat.TaskMetadata.DeviceName,
+		taskStat.TaskMetadata.ContainerPID, taskStat.TaskMetadata.NumberContainers)
+
+	returnError := false
+	for {
+		select {
+		case <-taskStat.Ctx.Done():
+			return nil
+		case err := <-errC:
+			seelog.Warnf("Error encountered processing metrics stream from host, this may affect "+
+				"cloudwatch metric accuracy: %s", err)
+			returnError = true
+		case rawStat, ok := <-awsvpcNetworkStats:
+			seelog.Info("got stats")
```

Comment:
i'd suggest not logging thing for every stats (when there's no error) since it can be noisy